### PR TITLE
Give global drivers the ability to hidetext whenever

### PR DESCRIPTION
**Pull request by AustinXII**
my compromise for this. http://www.smogon.com/forums/threads/a-hidetext-command-for-expired-user-objects.3609221/

### CODE REVIEW

**Comment by panpawn** [comment]
This should probably say "global drivers"

**Comment by panpawn** [comment]
... I meant the commit name >_>

**Comment by AustinXII** [comment]
Well it’s obvious now that the title’s changed :p

**Comment by Zarel** [comment]
No it's not?

**Comment by Zarel** [comment]
How does a title make a commit message obvious?

**Comment by Zarel** [comment]
You realize some people use Git, not GitHub, and can't see the pull request title?

**Comment by AustinXII** [comment]
@Zarel Oh well yeah I didn’t think that mattered, the merge commit takes the title does it not? I can change the actual commit name when I’m back on a computer if it matters though.

**Comment by panpawn** [comment]
>the merge commit takes the title does it not?

It doesn't. I mean, Zarel can change it when he squash merges it but why not just follow the guidelines outlined in [CONTRIBUTING.md](https://github.com/Zarel/Pokemon-Showdown/blob/master/CONTRIBUTING.md)?

**Comment by AustinXII** [comment]
@panpawn Well I thought I was until you pointed out I forgot global, I didn’t change the commit because I’m not on a computer but I thought changing the title would suffice

**Comment by Zarel** [comment]
They call it "squash and merge" because it's literally a squash and merge.

A squash by default just combines all the commits together into one. I'll usually manually edit the commit message to match the standards.

I normally don't care enough to tell the requester to update the commit message, but I do get kind of irritated if someone else does and the requester just brushes it off.

**Comment by Zarel** [comment]
Especially you, @AustinXII, you run the damagecalc so it's important that you actually understand the standards and why we have them.

**Comment by AustinXII** [comment]
@Zarel I understand the standards, it was just a miss typing on my part and I haven’t changed it just because I’m not near a computer. I thought this was “good enough” as it was just missing a global indentifier and I thought it would be changed in the actual important commit.

**Comment by AustinXII** [comment]
Also tbf I thought he only meant the title :P I didn’t try to just brush off panpawns concern.

**Comment by AustinXII** [comment]
@Zarel done

**Comment by Zarel** [comment]
Here, compromise: You can hidetext if the user is offline.

**Comment by AustinXII** [comment]
ooo I assumed you wouldn’t have wanted that for drivers, will update.

**Comment by AustinXII** [comment]
done @Zarel

**Comment by Zarel** [comment]
Also, here, use `room.log.hasUsername` as a typo guard:

7e26ea49df69b23db

File: chat-commands.js [hunk 1]
```diff
@@ -2377,19 +2377,22 @@ exports.commands = {
 	hidealtstext: 'hidetext',
 	hidetext: function (target, room, user, connection, cmd) {
 		if (!target) return this.parse('/help hidetext');
+		if (!room.log.hasUsername(target)) return false;
```
Comment (Zarel) [inline chat-commands.js]: No error message?

File: chat-commands.js [hunk 2]
```diff
@@ -2377,19 +2377,22 @@ exports.commands = {
...
 
 		this.splitTarget(target);
 		let targetUser = this.targetUser;
 		let name = this.targetUsername;
-		if (!targetUser) return this.errorReply("User '" + name + "' not found.");
-		let userid = targetUser.getLastId();
+		if (!targetUser && !user.can('lock')) return this.errorReply("User '" + name + "' not found.");
```
Comment (Zarel) [inline chat-commands.js]: Template string!

File: chat-commands.js [hunk 3]
```diff
@@ -2377,19 +2377,22 @@ exports.commands = {
...
+		let userid = targetUser ? targetUser.getLastId() : "";
```
Comment (Zarel) [inline chat-commands.js]: `''`

(CONTRIBUTING.md does talk about what string delimiters you should use in what situation, for the record)

Comment (Zarel) [reply chat-commands.js]: Also I hope you realize that your code doesn't actually work. At no point do you define `userid` correctly for offline users.

File: chat-commands.js [hunk 4]
```diff
@@ -2405,7 +2408,11 @@ exports.commands = {
...
 		} else {
-			this.addModAction("" + targetUser.name + "'s messages were cleared from  " + room.title + " by " + user.name + ".");
+			if (targetUser) {
+				this.addModAction(`${targetUser.name}'s messages were cleared from ${room.title} by ${user.name}.`);
+			} else {
+				this.addModAction(`${target}'s messages were cleared from ${room.title} by ${user.name}.`);
+			}
 			this.modlog('HIDETEXT', targetUser, null, {noip: 1, noalts: 1});
 			this.add('|unlink|' + hidetype + userid);
 			if (userid !== toId(this.inputUsername)) this.add('|unlink|' + hidetype + toId(this.inputUsername));
```
Comment (Zarel) [inline chat-commands.js]: Template string

File: chat-commands.js [hunk 5]
```diff
@@ -2405,7 +2408,11 @@ exports.commands = {
...
 			}
 		} else {
-			this.addModAction("" + targetUser.name + "'s messages were cleared from  " + room.title + " by " + user.name + ".");
+			if (targetUser) {
+				this.addModAction(`${targetUser.name}'s messages were cleared from ${room.title} by ${user.name}.`);
+			} else {
+				this.addModAction(`${target}'s messages were cleared from ${room.title} by ${user.name}.`);
+			}
 			this.modlog('HIDETEXT', targetUser, null, {noip: 1, noalts: 1});
 			this.add('|unlink|' + hidetype + userid);
```
Comment (Zarel) [inline chat-commands.js]: Template string

File: chat-commands.js [hunk 6]
```diff
@@ -2405,7 +2408,11 @@ exports.commands = {
 				this.add(`|unlink|${hidetype}${targetUser.prevNames[name]}`);
 			}
 		} else {
-			this.addModAction("" + targetUser.name + "'s messages were cleared from  " + room.title + " by " + user.name + ".");
+			if (targetUser) {
```
Comment (Zarel) [inline chat-commands.js]: There's no need for an `if` here, you can just use `${name}` for both.

File: chat-commands.js [hunk 7]
```diff
@@ -2405,7 +2408,11 @@ exports.commands = {
 				this.add(`|unlink|${hidetype}${targetUser.prevNames[name]}`);
```
Comment (Zarel) [inline chat-commands.js]: `name` should probably be renamed to `prevName` so it doesn't shadow `name`.

**Comment by Zarel** [comment]
Oh, that Travis error is my mistake; give me a moment.

**Comment by AustinXII** [comment]
sorry ;_;

File: chat-commands.js [hunk 8]
```diff
@@ -2403,10 +2406,14 @@ exports.commands = {
 				this.add(`|unlink|${hidetype}${targetUser.prevNames[prevName]}`);
 			}
 		} else {
-			this.addModAction(`${name}'s messages were cleared from ${room.title} by ${user.name}.`);
+			if (targetUser) {
```
Comment (Zarel) [inline chat-commands.js]: This still doesn't need to be changed. What was wrong with the old code? Can you show me the error message?

File: chat-commands.js [hunk 9]
```diff
@@ -2403,10 +2406,14 @@ exports.commands = {
...
+				this.addModAction(`${targetUser.name}'s messages were cleared from ${room.title} by ${user.name}.`);
+			} else {
+				this.addModAction(`${target}'s messages were cleared from ${room.title} by ${user.name}.`);
+			}
+			userid = toId(this.inputUsername);
```
Comment (Zarel) [inline chat-commands.js]: Why do you set `let userid = targetUser ? targetUser.getLastId() : "";` if you're just going to overwrite it here?

File: chat-commands.js [hunk 10]
```diff
@@ -2374,24 +2374,27 @@ exports.commands = {
...
+		} else if (!targetUser && user.can('lock')) {
 			hidetype = 'hide|';
 		} else {
-			return this.errorReply(`User "${name}" is not muted/banned from this room or locked.`);
+			return this.errorReply(`User ${name} is neither locked nor muted/banned from this room.`);
 		}
 
 		if (cmd === 'hidealtstext' || cmd === 'hidetextalts' || cmd === 'hidealttext') {
-			this.addModAction(`${name}'s alts' messages were cleared from ${room.title} by ${user.name}.`);
+			this.addModAction(`${targetUser.name}'s alts' messages were cleared from ${room.title} by ${user.name}.`);
```
Comment (Zarel) [inline chat-commands.js]: Ditto this. Why was this changed?

File: chat-commands.js [hunk 11]
```diff
@@ -2374,24 +2374,27 @@ exports.commands = {
 	hidealttext: 'hidetext',
 	hidealtstext: 'hidetext',
 	hidetext: function (target, room, user, connection, cmd) {
-		if (!target) return this.parse(`/help hidetext`);
+		if (!target) return this.parse('/help hidetext');
+		if (!room.log.hasUsername(target)) return this.errorReply(`User ${target} not found or has no roomlogs.`);
```
Comment (Zarel) [inline chat-commands.js]: This is, incidentally, the wrong place to check this.

Unless you want to make your code especially fancy, you should probably only check this if `targetUser` is not found.

File: chat-commands.js [hunk 12]
```diff
@@ -2374,24 +2374,27 @@ exports.commands = {
...
 		let name = this.targetUsername;
-		if (!targetUser) return this.errorReply(`User "${name}" not found.`);
-		let userid = targetUser.getLastId();
+		if (!targetUser && !user.can('lock')) return this.errorReply(`User ${name} not found.`);
+		let userid = targetUser ? targetUser.getLastId() : "";
 		let hidetype = '';
 		if (!user.can('mute', targetUser, room) && !this.can('ban', targetUser, room)) return;
 
-		if (targetUser.locked || Punishments.isRoomBanned(targetUser, room.id) || room.isMuted(targetUser) || user.can('rangeban')) {
+		if (targetUser && (targetUser.locked || Punishments.isRoomBanned(targetUser, room.id) || room.isMuted(targetUser) || user.can('lock'))) {
```
Comment (Zarel) [inline chat-commands.js]: At this line length, you should probably make

`(targetUser.locked || Punishments.isRoomBanned(targetUser, room.id) || room.isMuted(targetUser) || user.can('lock'))`

into its own variable

File: chat-commands.js (hunk 10 above)
Comment (Zarel) [reply chat-commands.js]: Unrelatedly, this crashes if `targetUser` doesn't exist. You should probably handle that case as well.

File: chat-commands.js (hunk 11 above)
Comment (AustinXII) [reply chat-commands.js]: checking this after the ``targetUser`` check doesn't work. I think this is the best place as the command should stop immediately if the target user has no lines in the chat.

Comment (Zarel) [reply chat-commands.js]: It's possible that the target user has no lines in the chat because they were using a different username.

Comment (AustinXII) [reply chat-commands.js]: Hmm I’ll have to hold off fixing this last thing for a few hours as I have some running around I have to do. Although thinking about it, wouldn’t it be better for ``hasUsername();`` check for alts?

Comment (panpawn) [reply chat-commands.js]: >checking this after the targetUser check doesn't work.

>wouldn’t it be better for hasUsername() check for alts?

... how would it do that for users with expired user objects?

Comment (AustinXII) [reply chat-commands.js]: Have you read the ``hasUsername()`` code?

Comment (AustinXII) [reply chat-commands.js]: Oh I guess my comment was sort of vague, I tried originally doing this after the ``!targetUser && user.can(‘lock’)`` check but doing something like ``!targetUser && !room.log.hasUsername(target)`` wasn’t working either

**Comment by AustinXII** [comment]
alright @Zarel everything besides the variable thing is done which for some reason broke the code. If you're fine with out it, this is done.

**Comment by Zarel** [comment]
I actually did explain in PM why the variable thing broke the code. It depends on `targetUser` existing, so the `targetUser &&` part can't actually be broken out.

**Comment by AustinXII** [comment]
Oh I think I was afk or I must have been leaving when you replied, I missed it mb
